lib/iris/coords.py

```diff
@@ -2775,6 +2779,366 @@ def xml_element(self, doc):
+        by the values in the connectivity's :attr:`indices` array.
+
+        """
+        return self._metadata_manager.cf_role.split("_")[1]
+
+    @property
+    def start_index(self):
+        """
+        The base value of the connectivity's :attr:`indices` array; either
+        ``0`` or ``1``.
+        **Read-only** - to change value: use :meth:`switch_start_index`.
+
+        """
+        return self._metadata_manager.start_index
+
+    @property
+    def element_dim(self):
+        """
+        The dimension of the connectivity's :attr:`indices` array that details
+        individual elements.
+        **Read-only** - to change value: use :meth:`switch_element_dim`.
+
+        """
+        return self._metadata_manager.element_dim
+
+    @property
+    def indices(self):
+        """
+        The index values describing the topological relationship of the
+        connectivity, as a NumPy array. Masked points indicate an element
+        shorter than the longest element described in this array - unused
+        index 'slots' are masked.
+        **Read-only** - index values are only meaningful when combined with
+        an appropriate :attr:`cf_role`, :attr:`start_index` and
+        :attr:`element_dim`. A new :class:`Connectivity` must therefore be
+        defined if different indices are needed.
+
+        """
+        return self._values
+
+    @property
+    def element_lengths(self):
+        """
+        A NumPy array of the lengths of each element in the :attr:`element_dim`
+        of the connectivity's :attr:`indices` array, accounting for masks if
+        present.
+
+        """
+        lengths = self._lazy_element_lengths(self.indices, self.element_dim)
+        return lengths.compute()
+
+    @property
+    def has_equal_element_lengths(self):
+        """
+        A boolean stating whether all values from
+        :attr:`element_lengths` are equal. Provided
+        for lazy calculation without needing to realise :attr:`indices` or
+        :attr:`element_lengths`.
+
+        """
+        lengths = self._lazy_element_lengths(self.indices, self.element_dim)
+        has_equal = lengths.min() == lengths.max()
+        return has_equal.compute()
+
+    def _validate_indices(self, indices):
+        def indices_error(message):
+            raise ValueError("Invalid indeces provided. " + message)
+
+        indices = self._sanitise_array(indices, 0)
+
+        indices_dtype = indices.dtype
+        if not np.issubdtype(indices_dtype, np.integer):
+            indices_error(
+                f"dtype must be numpy integer subtype, got: {indices_dtype} ."
+            )
+
+        indices_min = indices.min()
+        if _lazy.is_lazy_data(indices_min):
+            indices_min = indices_min.compute()
+        if indices_min < self.start_index:
+            indices_error(
+                f"Lowest index: {indices_min} < start_index: {self.start_index} ."
+            )
```

Comment:
I understand this validation step, but there is a cost here.

Firstly, this validation step is locked in, and there is absolutely no way (at the moment) for it to be avoided on the creation of a `Connectivity` instance.

Secondly, consider a `C1668` cubed-sphere quad mesh. That has 1668*1668*6 = 16,693,344 faces, each face with 4 nodes... so that's 66,773,376 node indexes to search through for the `min`. In terms of memory, you'd need a 4-byte integer to represent the maximum index value, so that's loading and searching through ~254MB.

This won't come close to even causing an issue, but this is all avoidable, along with all the lovely validation that follows this in this method.

Validation is good, but make it optional. The users may totally trust their data, given its provenance, and so validation is not necessary. Other users may definitely want to exercise this validation, but I'd argue that should be an explicit choice.

Reply:
I had imagined that lazy computation would take care of this problem, but given you raised a concern I wrote a test and it did indeed use around that much memory each time 😔. Given this I question whether it's worth going into the depth of checking variable face sizes during validation or whether we just validate the raw `shape` of the array.

Having disproved my assumptions about computation I'd actually question whether [`_lazy_element_lengths`](https://github.com/SciTools/iris/pull/3968/files/ff08a8c0fcd14955be960c91341fb6e55b492ff5#diff-8fd786735e6d5fbb391db7380cb986b8836869167159414ca5a83141860b835dR2884-R2889) is worth the resource in any case - i.e. is it worth giving `Connectivity` the intelligence to inspect any masks in `indices` to calculate individual element lengths? I'd consider removing all uses of it.

Reply:
Validate method should have a `shapes_only (bool)` parameter. If `shapes_only == True`, just validate the shapes, otherwise do the thorough validation that supports masked index 'slots'.

Reply:
Will keep the private method, which will be used during `__init__` validation using `shapes_only=True`. Will also offer a public method, which calls the private method always using `shapes_only=False`, since we can guarantee that the `Connectivity` instance has already been through the `shapes_only=True` validation and that `indices` have not been modified. Will be sure to warn of possible resource demand in the docstring.

Reply:
7354e89a998bd64bf8fff3166131628db385b2a4